### ISSUE TITLE
Nextcloud 'unable to connect' after disabling & enabling https (and changing certificates)

### ISSUE THREAD

**Issue by dinosmm** [nextcloud-snap/nextcloud-snap#1548]
#### Describe the bug
Please excuse the long-winded explanation, perhaps the background is important.
On Nextcloud (snap) 20.0.1, I tried to disable https so that I could set it up behind nginx. My intention was to have it work over http/2 to speed it up. The setup worked but it was very slow. I tried to undo everything I did (so, to set NC back to handling its own connection, enabling https), but it couldn't get a certificate, certbot connection keeps getting refused. Nothing wrong with firewalls or ports. NC then started being unable to connect. I changed ports etc., no luck. I set up the certificates manually, by copying the cert files to the directory required by enable-https, but it complained the certificate had expired (I had just got a new one manually on certbot, so that was baffling).
I tried lots of things, eventually I ended up reverting back to 19 via cli. 19 worked but again it was very slow, some elements (e.g. calendar) weren't loading. I 'refreshed' back to 20, and now it does not connect again. Curling localhost does not produce any output.

I have been using NC on snap for months flawlessly. At this point I am happy to compensate the time of anyone who can help solve this!

#### To Reproduce
As above, I made changes to ports, disabled/enabled https, and eventually reverted and refreshed.

#### Expected behavior
Nextcloud should work as before.

#### OS/snapd/snap version
Nextcloud 20.0.1snap1
Snap 2.47.1
Ubuntu 20.04

#### Logs
Too long to post, so I am attaching it.
[ncdebug.txt](https://github.com/nextcloud/nextcloud-snap/files/5543243/ncdebug.txt)

**Comment by kyrofa** [comment]
> Please excuse the long-winded explanation, perhaps the background is important.

Absolutely-- never apologize for giving more information than you think is required. I have a few ideas, but I need a little more info first.

1. You say you've reverted and are back up and running (albeit slowly). Are you currently using your nginx reverse proxy setup, then?
2. Is nginx on the same machine as the snap, or another machine? If the same machine, did you change the ports of the snap, then? If a different machine, did you update your DNS to point back at the snap when you wanted to move back? Did you wait for it to propagate, or is it possible that Let's Encrypt was still hitting the server when nginx was when checking the domain?

**Comment by kyrofa** [comment]
To completely disable HTTPS and get back to the point where you can get new certs, try this:

```
# Disable HTTPS
$ sudo nextcloud.disable-https

# Toast existing certs
$ sudo rm -rf /var/snap/nextcloud/current/certs
```

Now you're running unencrypted, but if you previously had a certificate from Let's Encrypt, the snap enabled HSTS which means your browser will _refuse to visit it_. You can obtain a fresh cert if you like, which will re-enable HSTS:

```
$ sudo nextcloud.enable-https lets-encrypt
```

**Comment by dinosmm** [comment]
First of all, thank you so much for helping out.

> 
>     1. You say you've reverted and are back up and running (albeit slowly). Are you currently using your nginx reverse proxy setup, then?

No. That was the intention, but when that did work, it was incredibly slow, so I then tried to backtrace all my steps to go back to the way it was, and that's when these problems started. Since then, I tried going back to using nginx again, but I couldn't even get it to do that.
At the moment, judging from netstat, the Nextcloud snap is not even listening for connections, whatever port I set on http or https.

>     2. Is nginx on the same machine as the snap, or another machine? If the same machine, did you change the ports of the snap, then? If a different machine, did you update your DNS to point back at the snap when you wanted to move back? Did you wait for it to propagate, or is it possible that Let's Encrypt was still hitting the server when nginx was when checking the domain?

Yes nginx is on the same machine as the snap. I just set different ports for nginx (I've always had Nextcloud on non-standard ports anyway).

> To completely disable HTTPS and get back to the point where you can get new certs, try this:

I did that too (I renamed certs to certs1 instead of deleting). 

> Now you're running unencrypted, but if you previously had a certificate from Let's Encrypt, the snap enabled HSTS which means your browser will refuse to visit it. You can obtain a fresh cert if you like, which will re-enable HSTS

I think that may be where the problem is then. When I reverted back to v19, Firefox told me my certificate was expired and wouldn't let me in because of HSTS. Now I'm back on 20 and it just says 'unable to connect', no HSTS warning or cert warning or anything. 
The problem though is that, I tried nextcloud.enable-https lets-encrypt - this won't verify, keeps saying 'connection refused' whatever I do. I am sure ports 80 and 443 are open and not being used by anything else (although happy to try whatever iteration of netstat might ultimately confirm this, I am going with netstat -lp). I have tried and can successfully create and renew certificates manually through certbot (outside the snap) for my domains, as this uses port 80 for verification I concluded the port is open and free, right?
And I tried creating certificates manually and running nextcloud.enable-https custom with the .pem files as parameters. On v19 I get a message saying certificates are installed and apache is restarted, but then FF says certificates are expired even though I just made them). On v20 I get 'certificates installed' but no 'apache restarted', and then FF says unable to connect.

**Comment by dinosmm** [comment]
This all happened last night. I just tried it again right now and it seems to be working. It's taken the new certificate I created manually (unless it renewed its own itself, not sure if there is a way to tell).
It's on its own https (not nginx), but still going very slow, and I have a large amount of transfers going on by it (my data directory is on GDrive). I also notice that some recent calendar events are missing, would the whole revert/refresh process have initiated a db restore?

I hope it eventually goes back to normal. If you have any insight as to what happened - why it was not working all last night and suddenly decided to work this morning, as well as any tips to optimise its speed, I would be most grateful.

**Comment by dinosmm** [comment]
I am just mentioning my findings as I go along, in case you have any idea what went on, as they keep getting weirder.
So I have lost several calendar events. I have several separate calendars - for one of them, I have lost recent events up to a week ago, however for another I have lost events going back 3 weeks. 
I looked at restoring a DB backup, however it seems that at some point in time a few months ago, NC decided not to back up oc_calendars tables... All I see in the backup list is oc_accounts, oc_activity, and for some backup points some additional tables relating to audio apps I don't even have installed any more.
I manually created a backup just now, and this one does indeed contain all tables including oc_calendars... I am now beyond baffled.

**Comment by dinosmm** [comment]
Now I am getting errors from the file sync client, such as: 

Broken webserver returning empty content length for non-empty file on resume <file path>

**Comment by dinosmm** [comment]
I believe my NC is being slow because it keeps downloading and uploading back and forth from/to GDrive (where the data directory & DB are stored). I left it running till now as I thought whatever it is doing it will finish eventually, but so far it's used 1.5TB and not slowing down, so I stopped the snap as I don't want it to use all my transfer quota.
I can see that it's transferring to Google servers and I can tell it's the GDrive mount, but I can't tell exactly what NC is actually doing - is there a way to find that out?

**Comment by kyrofa** [comment]
> is there a way to find that out?

You can try looking at the Nextcloud log (in `/var/snap/nextcloud/current/logs/nextcloud.log` if it's a newer instead, in your data directory if older), see if you see anything telltale in there.

**Comment by dinosmm** [comment]
Conclusion:
I found that NC could not write the log file where I told it to, due to permissions. It kept therefore writing a log file inside the data directory, which kept increasing in size with the error message about not being able to write the log file etc. I believe this was the cause of the increased server load and bandwidth usage.
In addition, once I restored the log file, I found that clicking on settings gave me an internal server error. Upon further investigation, the table oc_accounts did not exist in the db, so no user info could be found. 
I tried to restore recent db backups but the oc_accounts table seemed to be badly formatted in all of them - I kept getting "table oc_accounts does not contain a 'uid' column". The most recent backup that did restore was from September - this restored without any issues or error messages, however I lost one user whom I had added after that backup was taken.

I have no idea why the oc_accounts table would be gone from the db or why the log file situation arose. I can only assume this happened when I reverted back to 19 and then refreshed back to 20. Perhaps there are incompatibilities in the way the two versions set up the db?

Is there a way for the snap to check on this or to report if a db error occurs during revert/refresh?
